plugin/src/main/java/org/opensearch/sql/plugin/datasource/OpenSearchDataSourceMetadataStorage.java

```diff
@@ -96,28 +106,75 @@ public void createDataSourceMetadata(DataSourceMetadata dataSourceMetadata) {
     }
     IndexRequest indexRequest = new IndexRequest(DATASOURCE_INDEX_NAME);
     indexRequest.id(dataSourceMetadata.getName());
+    indexRequest.opType(DocWriteRequest.OpType.CREATE);
     ActionFuture<IndexResponse> indexResponseActionFuture;
+    IndexResponse indexResponse;
     try (ThreadContext.StoredContext storedContext = client.threadPool().getThreadContext()
         .stashContext()) {
       indexRequest.source(XContentParserUtils.convertToXContent(dataSourceMetadata));
       indexResponseActionFuture = client.index(indexRequest);
+      indexResponse = indexResponseActionFuture.actionGet();
+    } catch (VersionConflictEngineException exception) {
+      throw new IllegalArgumentException("A datasource already exists with name: "
+          + dataSourceMetadata.getName());
     } catch (Exception e) {
       throw new RuntimeException(e);
     }
-    IndexResponse indexResponse = indexResponseActionFuture.actionGet();
+
     if (indexResponse.getResult().equals(DocWriteResponse.Result.CREATED)) {
       LOG.debug("DatasourceMetadata : {}  successfully created", dataSourceMetadata.getName());
+    } else {
+      throw new RuntimeException("Saving dataSource metadata information failed with result : "
+          + indexResponse.getResult().getLowercase());
     }
   }
 
   @Override
   public void updateDataSourceMetadata(DataSourceMetadata dataSourceMetadata) {
```

Comment:
IMO again - side effect API should return status

Reply:
Any reference to best practices on side effect APIs. The calling methods would loose information on exception if I catch and return status.

Reply:
https://codereview.stackexchange.com/questions/11724/is-it-better-practice-to-have-void-method-throw-an-exception-or-to-have-the-meth
This questions discusses this.